nixos/modules/installer/cd-dvd/sd-image.nix

```diff
@@ -97,7 +96,7 @@ in
     system.build.sdImage = pkgs.stdenv.mkDerivation {
       name = config.sdImage.imageName;
 
-      buildInputs = with pkgs; [ dosfstools e2fsprogs mtools libfaketime utillinux ];
+      nativeBuildInputs = with pkgs.buildPackages; [ dosfstools e2fsprogs mtools libfaketime utillinux ];
```

Comment:
@dezgeg Actually, this `buildPackages` was necessary; not sure why. This isn't a `callPackage` expression, but I thought that using `pkgs.foo` in `nativeBuildInputs` would be the same as `pkgs.buildPackages.foo`. Yet, with the former you get the cross version.

Reply:
Yeah, exactly, derivations using `nativeBuildInputs` that aren't inside a `callPackage` are broken when cross-compiling and need this `nativeBuildInputs = with buildPackages; [ ... ];` dance.

I think the solution would be have something like `callPackagesInline` that instead of a path to a Nix expression takes the Nix expression directly. So this would become something like:

````
system.build.sdImage = pkgs.callPackageInline ({ stdenv, dosfstools, e2fsprogs, mtools, libfaketime, utillinux }: 
      name = config.sdImage.imageName;

      nativeBuildInputs = [ dosfstools e2fsprogs mtools libfaketime utillinux ];
      /* and so on */
}):

Reply:
@dezgeg `callPackage` can take a function instead of a file. But I'm not sure that's any more elegant than just being explicit...

Reply:
Ah, indeed `callPackage` can take a function. That can be used then without any changes to other places. Well, except the other places that do this `buildPackages` hack need to be fixed up to use `callPackage` as well.

> But I'm not sure that's any more elegant than just being explicit...

The fact that `nativeBuildInputs` works correctly in 99% of the cases (i.e. every package included from `all-packages.nix`) but _silently_ does the wrong thing when not inside `callPackage` is really, really terrible. If adding a `callPackage` to those 1% cases makes them work like the 99% cases then that's the correct action. And then `mkDerivation` should be changed to throw an error when cross-compiling and not inside `callPackage` instead of silently doing the wrong thing.

Reply:
That sounds like fighting a hack with a hack to me. The fact that callPackage is the solution here is a gross hack. There should be minimal relationship between mkDerivation and callPackage. callPackage just happened to be the easiest way to fix this for legacy code. To me it seems like the preference, if we'd started with cross compilation in mind, would have been to somehow be more explicit, not use this insane splicing that's hard to reason about. I don't think we're solving any problem by requiring the use of callPackage, we're just obscuring the code.

Reply:
> To me it seems like the preference, if we'd started with cross compilation in mind, would have been to somehow be more explicit, not use this insane splicing that's hard to reason about.

Yes, if someone had >10 years ago from the start decided that cross-compilation would be a thing to support and take it into the account in the design, things would have certainly been designed differently, no question about that. But hindsight is 20/20, and what we have now is what he have, and almost certainly is too late to change. Rewriting/redesigning interfaces for essentially a minority use case (and I really don't see the situation changing with the x86 monopoly for desktops and servers still being strong as ever) and breaking the existing code of the 90% of the userbase who have no need for cross compilation is not the correct tradeoff. (This is a personal opinion, but I don't think I'm the only one).

In other words, of the following choices:

a. Not supporting cross-compilation in Nixpkgs at all,
b. Supporting cross-compilation but with minimal possible changes to existing packages with compat tricks like splicing and separate `buildInputs`, `PATH` etc. logic in stdenv between native and cross cases,
c. Redesigning the entirety of nixpkgs around cross-compilation and changing each and single package,

I would pick b. any day, by a long shot (followed by a. then c.).

> I don't think we're solving any problem by requiring the use of callPackage, we're just obscuring the code.

Actually, regardless of cross-compilation, using `callPackage` here would be an improvement because then you could use `.override` to change say, the version of `e2fsprogs` used here. Currently when I try it I just get something like `error: attribute 'override' in selection path 'config.system.build.sdImage.override' not found` which is inconsistent with all the rest of the 99% of the derivations in Nixpkgs.

Reply:
Obviously I wasn't suggesting rewriting all of nixpkgs. I was saying that `callPackage` isn't holy in any way in this context. Native build input splicing isn't a great reason to use it when it's not necessary. In fact, I originally thought that splicing worked by mkDerivation checking the nativeBuildInputs and splicing those, not by splicing the package itself; that would have made a bit more sense to me, and would have made this fine.

Regardless, I hadn't thought about the desire to `override` this derivation normally. That's a good point. I've changed it to use `callPackage`, per your suggestion. Seems to work fine.